### PR TITLE
#67 - Added the ability to customize the store used for tracking rate limits

### PR DESCRIPTION
**Pull request by Svjard**
#### Checklist

- [X] run `npm run test` and `npm run benchmark`
- [X] tests and/or benchmarks are included
- [X] documentation is changed or added
- [X] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: example/example-custom.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+          trx
+            .raw(`INSERT INTO RateLimits(Source, Count, TTL) VALUES('${key}',1,${d.TTL || ttl}) ON CONFLICT(Source) DO UPDATE SET Count=Count+1`)
+            .then(() => {
+              cb(null, { current: d.Count ? d.Count + 1 : 1, ttl: d.TTL || ttl })
+            })
+            .catch(err => {
+              cb(err)
+            })
+        }
+      })
```
Comment (mcollina) [inline example/example-custom.js]: Can you add a catch handler?

File: index.js [hunk 2]
```diff
@@ -42,10 +42,15 @@ function rateLimitPlugin (fastify, settings, next) {
     whitelist: globalParams.whitelist
   }
 
-  if (settings.redis) {
-    pluginComponent.store = new RedisStore(settings.redis, 'fastify-rate-limit-', globalParams.timeWindow)
+  if (settings.store) {
+    const Store = settings.store
+    pluginComponent.store = new Store(globalParams.timeWindow, 'fastify-rate-limit-')
```
Comment (Eomm) [inline index.js]: Why this interface as a constructor?
the `fastify-rate-limit-` was an utill for redis.

What do you think about `globalParams` ?
So user can do a lot of (strange 😄) things as an increment doubled for users that are not in the whitelist

File: README.md [hunk 3]
```diff
@@ -77,6 +77,7 @@ fastify.register(require('fastify-rate-limit'), {
 - `whitelist`: array of string of ips to exclude from rate limiting. It can be a sync function with the signature `(req, key) => {}` where `req` is the Fastify request object and `key` is the value generated by the `keyGenerator`. If the function return a truthy value, the request will be excluded from the rate limit.
 - `redis`: by default this plugins uses an in-memory store, which is fast but if you application works on more than one server it is useless, since the data is store locally.<br>
 You can pass a Redis client here and magically the issue is solved. To achieve the maximum speed, this plugins requires the use of [`ioredis`](https://github.com/luin/ioredis)
+- `store`: a custom store to track requests and rates, for an example custom store please see the [`example/`](https://github.com/fastify/fastify-rate-limit/tree/master/example) folder
```
Comment (Eomm) [inline README.md]: Could we have a minimal store-example here?
The `CustomStore` in test is perfect

File: index.js (hunk 2 above)
Comment (Svjard) [reply index.js]: @Eomm Agree, that is a better call. I adjusted the PR accordingly and made the examples more explicit.

File: README.md (hunk 3 above)
Comment (Svjard) [reply README.md]: Added in the README an example as well as pointing them to the examples folder.

**Pull request merged**